kratos/linear_solvers/subspace_iteration_eigenvalue_solver.h

```diff
@@ -0,0 +1,615 @@
+
+            bool converged = true;
+            for (int i = 0; i< nroot; i++)
+            {
+                if (rtolv(i) > rtol) converged = false;
+            }
+            if (converged)
+            {
+                if (echo_level > 0) std::cout << "Convergence reached after " << nite << " iterations within a relative tolerance: " << rtol << std::endl;
+                break;
+            }
+            else{
+                if (nite >= nitem)
+                {
+                    if (echo_level > 0) std::cout << "Convergence not reached in " << nitem << " iterations." << std::endl;
+                    break;
+                }
+            }
+
+            for (int i = 0; i< nc ; i++)
+                prev_eigv(i) = eigv(i); // label 410 and 440
+
+            //TODO evaluate if necessary: GramSchmidt Orthogonalization
+        } while (true);
+
+        mpLinearSolver->FinalizeSolutionStep(rK, temp_tt, tt);
+
+        if(eigen_solver_successful)
+        {
+            // compute eigenvectors
+            for (int j = 0; j< nc; j++)
+            {
+                for (int k = 0; k< nn; k++)
+                    tt(k) = r(k,j);
+
+                // K*temp_tt = tt
+                mpLinearSolver->PerformSolutionStep(rK, temp_tt, tt);
+
+                for (int k = 0; k< nn; k++)
+                    r(k,j) = temp_tt(k);
+            }
+
+            // copy results to function parameters
+            if (static_cast<int>(rEigenvalues.size()) != nroot)
+                rEigenvalues.resize(nroot);
+            if (static_cast<int>(rEigenvectors.size1()) != nroot || static_cast<int>(rEigenvectors.size2()) != nn)
+                rEigenvectors.resize(nroot, nn);
+
+            for (int i = 0; i< nroot; i++)
+            {
+                rEigenvalues(i) = eigv(i);
+                for (int j=0; j< nn; j++)
+                    rEigenvectors(i,j) = r(j,i);
+            }
+
+            if (echo_level > 0) KRATOS_WATCH(rEigenvalues);
+
+            // make vectors M-normalized
+            if (mParam["mass_normalization"].GetBool())
+            {
+                MassNormalizeEigenVectors(rM, rEigenvectors);
+            }
+
+            // orient vectors
+            if (mParam["orient_eigen_vectors"].GetBool()){
+                OrientEigenVectors(rEigenvectors);
+            }
+            // TODO
+            // 1. speed up
+            // 2. sturm sequence check
+        }
+        else{
+            KRATOS_ERROR << "Solution failed!" <<std::endl;
+        }
+
+        if (echo_level > 0) {
+            std::cout  << "SubspaceIterationEigenvalueSolver completed in " <<
+            OpenMPUtils::GetCurrentTime() - start_time << " seconds" << std::endl;
+        }
+
+        return;
+    }
+
+    /**
+     * This method returns directly the first eigen value obtained
+     * @param rK: The stiffness matrix
+     * @param rM: The mass matrix
+     * @return The first eigenvalue
+     */
+    double GetEigenValue(
```

Comment:
can you make this more general? Such that you can get any eigenvalue (in case it exists).
Then you could have the first eigenvalue as default argument

Reply:
this function is currently used only by the tests.
Do you happen to know how to deal with default values in a function exposed to python?

Reply:
I use that [here](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/MappingApplication/custom_python/add_custom_utilities_to_python.cpp)

Reply:
To make it even more general, i think it makes more sense to expose the Solve function:

```python
            K = KratosMultiphysics.CompressedMatrix(n,n)
            M = KratosMultiphysics.CompressedMatrix(n, n)
            vec = KratosMultiphysics.Matrix(0, 0)
            val = KratosMultiphysics.Vector(0)
            eigen_solver.Solve(K, M, val, vec)
```